staging/src/k8s.io/client-go/transport/round_trippers.go

```diff
@@ -422,3 +422,130 @@ func (rt *debuggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, e
 func (rt *debuggingRoundTripper) WrappedRoundTripper() http.RoundTripper {
 	return rt.delegatedRoundTripper
 }
+
+func isLegalHeaderKey(key string) bool {
+	for i := 0; i < len(key); i++ {
+		if !legalHeaderByte(key[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func legalHeaderByte(b byte) bool {
+	return int(b) < len(legalHeaderKeyBytes) && legalHeaderKeyBytes[b]
+}
+
+func shouldEscape(b byte) bool {
+	// url.PathUnescape() returns an error if any '%' is not followed by two
+	// hexadecimal digits, so we'll intentionally encode it.
+	return !legalHeaderByte(b) || b == '%'
+}
+
+func headerKeyEscapedLen(s string) int {
+	encodedLen := len(s)
+	for i := 0; i < len(s); i++ {
+		if shouldEscape(s[i]) {
+			encodedLen += 2
+		}
+	}
+	return encodedLen
+}
+
+func headerKeyEscape(key string) string {
+	buf := strings.Builder{}
+	buf.Grow(headerKeyEscapedLen(key))
```

Comment:
You can just write to buf without pre-scanning key to allocate memory

Reply:
I realize that this is a micro-optimization, but it was done in response to feedback earlier about repetitive string concatenations being malloc-heavy.